### PR TITLE
Unix socket doc

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @frewsxcv

### CODE REVIEW

File: src/libstd/sys/unix/ext/net.rs [hunk 1]
```diff
@@ -85,6 +85,12 @@ enum AddressKind<'a> {
 }
 
 /// An address associated with a Unix socket.
+///
+/// # Examples
+///
+/// ```
+///
+/// ```
```
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: This example doesn't seem particularly helpful :)

File: src/libstd/sys/unix/ext/net.rs [hunk 2]
```diff
@@ -121,6 +127,22 @@ impl SocketAddr {
     }
 
     /// Returns true if and only if the address is unnamed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::os::unix::net::UnixListener;
+    ///
+    /// let socket = match UnixListener::bind("/tmp/sock") {
```
Comment (nagisa) [inline src/libstd/sys/unix/ext/net.rs]: IMHO every single example exhaustively matching is sort of overzealous.

```
if let Ok(socket) = UnixListener::bind("/tmp/sock") {
     // show off the feature
}
```

would be a much better alternative IMO.

File: src/libstd/sys/unix/ext/net.rs [hunk 3]
```diff
@@ -212,6 +251,20 @@ impl fmt::Debug for UnixStream {
 
 impl UnixStream {
     /// Connects to the socket named by `path`.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::os::unix::net::UnixStream;
+    ///
+    /// let socket = match UnixStream::connect("/tmp/sock") {
```
Comment (nagisa) [inline src/libstd/sys/unix/ext/net.rs]: unused variable :)

File: src/libstd/sys/unix/ext/net.rs [hunk 4]
```diff
@@ -229,6 +282,20 @@ impl UnixStream {
...
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::os::unix::net::UnixStream;
+    ///
+    /// let (sock1, sock2) = match UnixStream::pair() {
+    ///     Ok((sock1, sock2)) => (sock1, sock2),
+    ///     Err(e) => {
+    ///         println!("Couldn't connect: {:?}", e);
```
Comment (nagisa) [inline src/libstd/sys/unix/ext/net.rs]: Couldn’t *pair*.

File: src/libstd/sys/unix/ext/net.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/libstd/sys/unix/ext/net.rs]: Oh damn!

File: src/libstd/sys/unix/ext/net.rs (hunk 3 above)
Comment (GuillaumeGomez) [inline src/libstd/sys/unix/ext/net.rs]: I don't see it as an issue here, on the opposite. The point is to show people how it works, and creating a (unused) variable helps it.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by GuillaumeGomez** [comment]
And urls fixed as well. Build fails but doesn't have errors so I guess it's okay?

File: src/libstd/sys/unix/ext/net.rs [hunk 5]
```diff
@@ -241,62 +305,159 @@ impl UnixStream {
...
 
     /// Moves the socket into or out of nonblocking mode.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::os::unix::net::UnixStream;
+    ///
+    /// let socket = UnixStream::connect("/tmp/sock").unwrap();
+    /// socket.set_nonblocking(true).expect("Couldn't set non blocking");
```
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: Nit: 'nonblocking' is one word

File: src/libstd/sys/unix/ext/net.rs [hunk 6]
```diff
@@ -692,6 +1104,15 @@ impl UnixDatagram {
...
     ///
     /// On success, returns the number of bytes written.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::os::unix::net::UnixDatagram;
+    ///
+    /// let sock = UnixDatagram::unbound().unwrap();
+    /// sock.send_to(b"omelette AU fromage", "/some/sock").expect("send_to function failed");
```
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: Nit: why is "AU" capitalized?

File: src/libstd/sys/unix/ext/net.rs [hunk 7]
```diff
@@ -716,50 +1137,134 @@ impl UnixDatagram {
...
     /// On success, returns the number of bytes written.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::os::unix::net::UnixDatagram;
+    ///
+    /// let sock = UnixDatagram::unbound().unwrap();
+    /// sock.connect("/some/sock").expect("Couldn't connect");
+    /// sock.send(b"omelette AU fromage").expect("send_to function failed");
```
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: Ditto "AU"

File: src/libstd/sys/unix/ext/net.rs [hunk 8]
```diff
@@ -121,6 +136,16 @@ impl SocketAddr {
...
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::os::unix::net::UnixListener;
+    ///
+    /// let socket = UnixListener::bind("/tmp/sock").unwrap();
+    /// let addr = socket.local_addr().expect("Couldn't get local address");
+    /// assert_eq!(addr.is_unnamed(), false);
+    /// ```
```
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: Can we add an example here where `is_unnamed()` is `true`?

File: src/libstd/sys/unix/ext/net.rs [hunk 9]
```diff
@@ -131,6 +156,17 @@ impl SocketAddr {
...
+    /// # Examples
+    ///
+    /// ```
+    /// use std::os::unix::net::UnixListener;
+    /// use std::path::Path;
+    ///
+    /// let socket = UnixListener::bind("/tmp/sock").unwrap();
+    /// let addr = socket.local_addr().expect("Couldn't get local address");
+    /// assert_eq!(addr.as_pathname(), Some(Path::new("/tmp/sock")));
+    /// ```
```
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: Same here. If we can create an unnamed socket, can we show here that `as_pathname()` is `None`?

**Review by frewsxcv — COMMENTED** [review]
Looks good, just a few comments 🎊

File: src/libstd/sys/unix/ext/net.rs (hunk 6 above)
Comment (GuillaumeGomez) [inline src/libstd/sys/unix/ext/net.rs]: To help english people improve their french. :p

Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: For me, a person who doesn't know any French, it doesn't help me learn French, it just makes me confused :)

Comment (GuillaumeGomez) [inline src/libstd/sys/unix/ext/net.rs]: Haha. Well, I'll "fixed" it then.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/libstd/sys/unix/ext/net.rs [hunk 10]
```diff
@@ -121,6 +136,21 @@ impl SocketAddr {
...
+    /// # Examples
+    ///
+    /// ```
+    /// use std::os::unix::net::{UnixDatagram, UnixListener};
+    ///
+    /// let socket = UnixListener::bind("/tmp/sock").unwrap();
+    /// let addr = socket.local_addr().expect("Couldn't get local address");
+    /// assert_eq!(addr.is_unnamed(), false);
+    ///
+    /// // with is_unnamed() as true now:
```
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: Nit: Alternatively, you could also do something like this:

    # Examples

    An unnamed address:

    ```
    ...
    ```

    A named address:

    ```
    ...
    ```

File: src/libstd/sys/unix/ext/net.rs [hunk 11]
```diff
@@ -241,62 +315,159 @@ impl UnixStream {
...
     /// data, and options set on one stream will be propogated to the other
     /// stream.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::os::unix::net::UnixStream;
+    ///
+    /// let socket = UnixStream::connect("/tmp/sock").unwrap();
+    /// let sock_copy = socket.try_clone().expect("Couldn't clone socket...");
```
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: Nit: The ellipsis here doesn't seem necessary.

**Review by frewsxcv — APPROVED** [review]
r=me unless you want to address the nits

File: src/libstd/sys/unix/ext/net.rs (hunk 11 above)
Comment (GuillaumeGomez) [inline src/libstd/sys/unix/ext/net.rs]: The ellipsis? What do you mean? The `expect` call?

File: src/libstd/sys/unix/ext/net.rs (hunk 10 above)
Comment (GuillaumeGomez) [inline src/libstd/sys/unix/ext/net.rs]: Oh good point!

File: src/libstd/sys/unix/ext/net.rs (hunk 11 above)
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: `...`

https://en.wikipedia.org/wiki/Ellipsis

**Comment by GuillaumeGomez** [comment]
I addressed the nits.

File: src/libstd/sys/unix/ext/net.rs [hunk 12]
```diff
@@ -131,6 +168,22 @@ impl SocketAddr {
...
+    ///
+    /// ```
+    /// use std::os::unix::net::{UnixDatagram, UnixListener};
+    /// use std::path::Path;
+    ///
+    /// let socket = UnixListener::bind("/tmp/sock").unwrap();
+    /// let addr = socket.local_addr().expect("Couldn't get local address");
+    /// assert_eq!(addr.as_pathname(), Some(Path::new("/tmp/sock")));
+    ///
+    /// // with as_pathname() as true now:
```
Comment (frewsxcv) [inline src/libstd/sys/unix/ext/net.rs]: Did you want to do the same thing here?

**Comment by frewsxcv** [comment]
r=me unless you want to address that comment

**Comment by GuillaumeGomez** [comment]
And updated too.

**Comment by frewsxcv** [comment]
@bors r-

Errors: https://github.com/rust-lang/rust/pull/38387#issuecomment-267483811

**Comment by GuillaumeGomez** [comment]
And fixed. I re-r+ if you don't mind @frewsxcv.

@bors: r=frewsxcv

**Pull request merged**